controller/src/main/java/io/pravega/controller/store/stream/PravegaTablesReaderGroup.java

```diff
@@ -52,132 +53,131 @@
     private static final String CONFIGURATION_KEY = "configuration";
     private static final String STATE_KEY = "state";
 
+    private static final TagLogger log = new TagLogger(LoggerFactory.getLogger(PravegaTablesReaderGroup.class));
+
     private final PravegaTablesStoreHelper storeHelper;
-    private final Supplier<CompletableFuture<String>> readerGroupsInScopeTableNameSupplier;
-    private AtomicReference<String> idRef;
-    private final ScheduledExecutorService executor;
+    private final BiFunction<Boolean, OperationContext, CompletableFuture<String>> readerGroupsInScopeTableNameSupplier;
+    private final AtomicReference<String> idRef;
 
     @VisibleForTesting
     PravegaTablesReaderGroup(final String scopeName, final String rgName, PravegaTablesStoreHelper storeHelper,
-                        Supplier<CompletableFuture<String>> rgInScopeTableNameSupplier,
-                        ScheduledExecutorService executor) {
+                             BiFunction<Boolean, OperationContext, CompletableFuture<String>> rgInScopeTableNameSupplier,
+                             ScheduledExecutorService executor) {
         super(scopeName, rgName);
         this.storeHelper = storeHelper;
         this.readerGroupsInScopeTableNameSupplier = rgInScopeTableNameSupplier;
         this.idRef = new AtomicReference<>(null);
-        this.executor = executor;
     }
 
-    private CompletableFuture<String> getId() {
+    private CompletableFuture<String> getId(OperationContext context) {
         String id = idRef.get();
         if (!Strings.isNullOrEmpty(id)) {
             return CompletableFuture.completedFuture(id);
         } else {
-            return readerGroupsInScopeTableNameSupplier.get()
-                    .thenCompose(readerGroupsInScopeTable ->
-                            storeHelper.getEntry(readerGroupsInScopeTable, getName(),
-                                    x -> BitConverter.readUUID(x, 0)))
+            // first get the scope id from the cache.
+            // if the cache does not contain scope id then we load it from the supplier. 
+            // if cache contains the scope id then we load the streamid. if not found, we load the whole shit
+            return Futures.exceptionallyComposeExpecting(
```

Comment:
you are right and its triplicated actually. 
but in three different classes that are otherwise unrelated. 
let me see if i can take this into some common helper.

Reply:
done